kong/dns/client.lua

```diff
@@ -0,0 +1,761 @@
+      end
+    end,
+
+    -- @channel: event channel name, such as "mlcache:invalidate:dns_cache"
+    -- @data: mlcache's key name, such as "<qname>:<qtype>"
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc             = ipc,
+    neg_ttl         = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    lru_size        = opts.cache_size or 10000,
+    resty_lock_opts = resty_lock_opts,
+  })
+
+  if not cache then
+    return nil, "could not create mlcache: " .. err
+  end
+
+  if opts.cache_purge then
+    cache:purge(true)
+  end
+
+  -- parse order
+  if opts.order and table_isempty(opts.order) then
+    return nil, "Invalid order array: empty record types"
+  end
+
+  local order = opts.order or DEFAULT_ORDER
+  local search_types = {}
+  local preferred_ip_type
+
+  for i, typstr in ipairs(order) do
+    local qtype = NAME_TO_TYPE[typstr:upper()]
+    if not qtype then
+      return nil, "Invalid dns record type in order array: " .. typstr
+    end
+
+    search_types[i] = qtype
+
+    if (qtype == TYPE_A or qtype == TYPE_AAAA) and not preferred_ip_type then
+      preferred_ip_type = qtype
+    end
+  end
+
+  preferred_ip_type = preferred_ip_type or TYPE_A
+
+  -- parse hosts
+  local hosts = init_hosts(cache, opts.hosts, preferred_ip_type)
+
+  return setmetatable({
+    cache         = cache,
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    search_types  = search_types,
+    -- quickly accessible constant empty answers
+    EMPTY_ANSWERS = tablex_readonly({
+      errcode = EMPTY_RECORD_ERROR_CODE,
+      errstr  = EMPTY_RECORD_ERROR_MESSAGE,
+      ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    }),
+  }, mt)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)  -- 0xffffffff for maximum TTL value
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+
+    if answer.type == TYPE_CNAME then
```

Comment:
cache this var, `local answer_type = answer.type`

Reply:
done https://github.com/Kong/kong/pull/12305/commits/b87cc14ab796307cf88d07c7aed08475a9bb0497